### DISCUSSION TITLE
[Feature request] Dynamic Interfaces

### DISCUSSION THREAD

**Discussion by MHanafy** [dotnet/csharplang#2290] [Language Ideas]
When working on large codebases, sometimes there's a need to work on a large set of classes that all shares a common set of properties.
The typical solution is, to create a new interface, and make all these classes implement it; then refer to them through the interface.
While this's sometimes doable, the larger the codebase the harder it's to do.

I'm thinking about a simple solution, where C# allows a dynamic interface; then when using type comparison/casting to a dynamic interface; the class simply needs to match the interface members without a need to change the original class.

This will be very useful in maintaining large, long running projects that has a lot of classes.

```c#
class Supplier{
    public string Name {get; set;}
    public string EMail {get; set;}
    //--- Other irrelevant attributes
}

class Employee{
    public string Name {get; set;}
    public string EMail {get; set;}
    //--- Other irrelevant attributes
}

class MissingProps{
    public string Name {get; set;}
    //Note that EMail is missing
}

dynamic interface IContact{
    string Name {get; set;}
    string EMail {get; set;}
}

public static void Main(){
    var supplier = new Supplier{Name = "SomeName", EMail = "Mail@Provider.com" }
    var employee = new Employee{Name = "AnotherName", EMail = "Mail@Provider.com" }
    var missingProps= new MissingProps{Name = "ThirdName" }
    // below calls are successful because interface resolution is done dynamically,
    // since Supplier and Employee has the right signature
    var iSuppContact = supplier as IContact; 
    var iEmpContact = (IContact) supplier;

    //below call fails because of missing properties
    var iMissingProps = (IContact) missingProps;
}
```

**Comment by HaloFour** [comment]
I think #110 (shapes) would cover this.

**Comment by svick** [comment]
I don't see why this needs to be a language feature. You can use a library like like [ImpromptuInterface](https://github.com/ekonbenefits/impromptu-interface) instead:

```c#
var iSuppContact = supplier.ActLike<IContact>();
```

ImpromptuInterface does not throw an exception from `ActLike` when a member is missing, instead it throws when you attempt to access that member. But you could write a library that behaves the way you want.

**Comment by MHanafy** [comment]
@svick Yes, this approach can be used; the benefit of having this as a language feature is having compile time checks; the code won't compile until the classes matches the interface.

**Comment by svick** [comment]
@MHanafy That sounds like something that could be done with a Roslyn analyzer.

**Comment by Joe4evr** [comment]
> the benefit of having this as a language feature is having compile time checks; the code won't compile until the classes matches the interface.

That's exactly what Shapes are all about.
